yt_dlp/extractor/crunchyroll.py

```diff
@@ -135,47 +145,30 @@ def _call_api(self, path, internal_id, lang, note='api', query={}):
-            if stream_type not in requested_formats:
-                continue
-            for stream in traverse_obj(streams, lambda _, v: v['url']):
-                hardsub_lang = stream.get('hardsub_locale') or ''
-                format_id = join_nonempty(stream_type, format_field(stream, 'hardsub_locale', 'hardsub-%s'))
-                available_formats[hardsub_lang] = (stream_type, format_id, hardsub_lang, stream['url'])
+        available_formats = {'': ('', '', stream_response['url'])}
+        for hardsub_lang, stream in traverse_obj(stream_response, ('hardSubs', {dict.items}, ...)):
+            available_formats[hardsub_lang] = (f'hardsub-{hardsub_lang}', hardsub_lang, stream['url'])
 
         requested_hardsubs = [('' if val == 'none' else val) for val in (self._configuration_arg('hardsub') or ['none'])]
-        if '' in available_formats and 'all' not in requested_hardsubs:
-            full_format_langs = set(requested_hardsubs)
-            self.to_screen(
-                'To get all formats of a hardsub language, use '
-                '"--extractor-args crunchyrollbeta:hardsub=<language_code or all>". '
-                'See https://github.com/yt-dlp/yt-dlp#crunchyrollbeta-crunchyroll for more info',
-                only_once=True)
-        else:
-            full_format_langs = set(map(str.lower, available_formats))
 
-        audio_locale = traverse_obj(stream_response, ((None, 'meta'), 'audio_locale'), get_all=False)
+        audio_locale = stream_response.get('audioLocale')
         hardsub_preference = qualities(requested_hardsubs[::-1])
         formats = []
-        for stream_type, format_id, hardsub_lang, stream_url in available_formats.values():
-            if stream_type.endswith('hls'):
-                if hardsub_lang.lower() in full_format_langs:
-                    adaptive_formats = self._extract_m3u8_formats(
-                        stream_url, display_id, 'mp4', m3u8_id=format_id,
-                        fatal=False, note=f'Downloading {format_id} HLS manifest')
-                else:
-                    adaptive_formats = (self._m3u8_meta_format(stream_url, ext='mp4', m3u8_id=format_id),)
-            elif stream_type.endswith('dash'):
-                adaptive_formats = self._extract_mpd_formats(
-                    stream_url, display_id, mpd_id=format_id,
-                    fatal=False, note=f'Downloading {format_id} MPD manifest')
-            else:
-                self.report_warning(f'Encountered unknown stream_type: {stream_type!r}', display_id, only_once=True)
-                continue
+        for format_id, hardsub_lang, stream_url in available_formats.values():
+            adaptive_formats = self._extract_mpd_formats(
+                stream_url, display_id, mpd_id=format_id, headers=CrunchyrollBaseIE._AUTH_HEADERS,
+                fatal=False, note=f'Downloading {f"hardsub {hardsub_lang} " if hardsub_lang else ""}MPD manifest')
```

Comment:
This should employ the same lazy evaluation that hls had before. see [`_m3u8_meta_format` in `ie/common.py:`](<https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/common.py#L1985-L1995>) for how to implement the dash version using `http_dash_segments`. For this the above removed block has to be restored.

I like showing the hardsub language, but I would like to also have the `format_id`. How about something like this for that:
```suggestion
            if hardsub_lang.lower() in full_format_langs:
                adaptive_formats = self._extract_mpd_formats(
                    stream_url, display_id, mpd_id=format_id, headers=CrunchyrollBaseIE._AUTH_HEADERS,
                    fatal=False, note=f'Downloading {format_id} {f" ({hardsub_lang})" if hardsub_lang else ""}MPD manifest')
            else:
                adaptive_formats = [{
                    ...
                }]
```

Reply:
Good point. Should I add the dict in-place or create a separate method for it (just like `_meta_m3u8_format` is), and how about the extension, should it just be `None`?

Reply:
add the dict in place. `_m3u8_meta_format` was only ever used in crunchyroll (I found it by accident) so creating a specific functino for it seems overkill